bell.py

```diff
@@ -24,12 +34,69 @@ def bell(webhook_url):
     response = ec2.describe_instance_status(InstanceIds=[instance_id])
     status = response["InstanceStatuses"][0]["InstanceState"]["Name"]
     logger.debug(f"ec2 instance status: {status}")
+    return f":bellhop_bell: {ec2_instance_name} is {status}"
+
+
+def send_slack_message(webhook_url, message):
+    logger.debug(f"slack message: {message}")
+
+    data = json.dumps({"text": message}).encode("utf-8")
+
+    request = urllib.request.Request(webhook_url)
+    request.add_header("Content-Type", "application/json; charset=utf-8")
+    request.add_header("Content-Length", len(data))
+
+    try:
+        response = urllib.request.urlopen(request, data, timeout=5)
+        logger.debug(f"slack response code: {response.getcode()}")
+    except urllib.request.URLError:
+        logger.error(
+            "slack message not sent. probably timeout. this is normal if you are testing or running locally."
```

Comment:
Looks like this isn't catching all errors; when i try to run locally I get:

```
$ bell $WEBHOOKURL                                                                          
Traceback (most recent call last):
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/util/connection.py", line 95, in create_connection
    raise err
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/util/connection.py", line 85, in create_connection
    sock.connect(sa)
OSError: [Errno 113] No route to host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 398, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connection.py", line 239, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1011, in _send_output
    self.send(msg)
  File "/usr/lib/python3.8/http/client.py", line 951, in send
    self.connect()
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fd027499250>: Failed to establish a new connection: [Errno 113] No route to host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/adapters.py", line 440, in send
    resp = conn.urlopen(
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 785, in urlopen
    retries = retries.increment(
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='169.254.169.254', port=80): Max retries exceeded with url: /latest/meta-data/instance-id (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd027499250>: Failed to establish a new connection: [Errno 113] No route to host'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/bin/bell", line 8, in <module>
    sys.exit(app())
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/typer/main.py", line 500, in wrapper
    return callback(**use_params)  # type: ignore
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/bell.py", line 12, in bell
    response = requests.get("http://169.254.169.254/latest/meta-data/instance-id")
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/sessions.py", line 529, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/sessions.py", line 645, in send
    r = adapter.send(request, **kwargs)
  File "/home/matthew/Documents/treeconomy/annotation/build/virtualenv/lib/python3.8/site-packages/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='169.254.169.254', port=80): Max retries exceeded with url: /latest/meta-data/instance-id (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd027499250>: Failed to establish a new connection: [Errno 113] No route to host'))

Reply:
This seems to be using an earlier version though as the error comes from `requests`. Are you sure you are on the latest version? You might need to force it to reinstall as I did not upgrade the version, maybe I should?

Reply:
Note that in the latest version the syntax for sending an alert changed slightly to allow for the new functionality. So to test locally you can do `bell --webhook-url SLACK_WEBHOOK_URL` or setup an environment variable with that name.